docs/decisions/0003-lightweight-extension-points.rst

```diff
@@ -0,0 +1,202 @@
+
+- Modify the context passed to ``legacy/response/oa_response.html`` 
+- Change the template that is rendered to the student
+- Send students' submission data to another service
+
+Let's say you want to add an acknowledgment notice to your submission template so students know their information is being shared with third-party services when submitting a response. The extension developer could implement a `pipeline step`_ for the filter that changes the ``oa_response.html`` template for an ``oa_response_ack_modified.html`` template with its context:
+
+
+..  code-block:: python
+
+    from openedx_filters import PipelineStep
+    
+    
+    class ORASubmissionViewAcknowledgeWarning(PipelineStep):
+        """Add warning message about sharing users' information to the ORA submission view."""
+    
+        def run_filter(  # pylint: disable=unused-argument, disable=arguments-differ
+            self, context: dict, template_name: str
+        ) -> dict:
+            """
+            Execute filter that loads the submission template with a warning message that
+            notifies the user that the submission will be sent to a 3rd party service.
+    
+            Args:
+                context (dict): The context dictionary.
+                template_name (str): ORA template name.
+    
+            Returns:
+                dict: The context dictionary and the template name.
+            """
+            return {
+                "context": context,
+                "template_name": "some_plugin/oa_response_with_acknowledgement.html",
+            }
+
+See `how to implement pipeline steps`_ for more information. Now, by listening to the `Open edX Event`_, the developer could act on the submission-created notification. Since the event payload has enough information to get the student's submissions, including files, the event receiver can obtain the submission to send it to another service for analysis:
+
+
+..  code-block:: python
+
+    from some_plugin.tasks import ora_submission_created_processing_task
+
+    @receiver(ORA_SUBMISSION_CREATED)
+    def ora_submission_created(submission, **kwargs):
+        """
+        Handle the ORA_SUBMISSION_CREATED event.
+    
+        Args:
+            submission (ORASubmissionData): The ORA submission data.
+        """
+        ora_submission_created_processing_task.delay(
+            submission.id,
+            submission.file_downloads,
+        )
+
+See `how to listen for Open edX Events`_ for more information. 
+
+Extension developers could interact with an essential part of the student's assessment lifecycle with these changes. But when none of these extension points are configured for use, then ORA assessments will behave as usual.
+
+Rejected Alternatives
```

Comment:
Given that there is currently no other option for extending ORA without a fork we are not really rejecting any other alternative at this point. It could be argued that we are rejecting the construction of a larger (or more comprehensive) framework for extension but I don't see it like that. It is more like this is the first step towards a larger framework. If we were to propose a project to extend ORA with a mechanism for dependency injection we would still propose it to be build on top of the hooks framework.

At this ADR we are only commiting to the first few hooks because we understand very well the effort that it requires. However there is no technical limit for this proposal to grow into more hooks and to eventually support a broad array of extension use cases. Maybe this reflection is more for the consequences part. We are opening the door to more extension points in the future that are also built using events and filters and we can make it more explicit.

Reply:
I wasn't sure about what to put here, so I'll use this comment instead. Thanks.

Reply:
I suppose an alternative approach would be to make adding grading steps much more robust, using inversion of control and dependency injection to add steps rather than forks.  That would likely be significantly more work and would add another divergent pattern to events which we are using elsewhere.  A couple of general questions.  Could we imagine a solution where an event could be added to a function without changing the code?  In other languages than python you might do this with aspects, AOP point cuts or similar.  Not sure this is a good idea, so not advocating for it currently, but it would be extremely flexible.

Reply:
> Could we imagine a solution where an event could be added to a function without changing the code?

That's an approach that we haven't explored yet. The closest example I've seen is how people have implemented filters as decorators when used in multiple places (here's an example: https://github.com/openedx/event-routing-backends/blob/master/event_routing_backends/processors/openedx_filters/decorators.py#L9). This approach looks cleaner and is easier to maintain than duplicating code. However, I think we might be missing the explicitness of what's been processed and where for more complex functions. I imagine there's a way to solve that. It would be interesting to look further into it!